File: reports/20260826-173422/results/storage/data-plane/js-ts/crud/storage-dp-js-ts-encrypted-uploader/js-ts-azure-tools/baseline/generated-code/src/keyManagement.ts
import { randomBytes } from "node:crypto";

import type { TokenCredential } from "@azure/core-auth";
import {
  CryptographyClient,
  KeyClient,
  KnownEncryptionAlgorithms,
} from "@azure/keyvault-keys";

const DATA_KEY_LENGTH_BYTES = 32;
const KEY_WRAP_ALGORITHM = KnownEncryptionAlgorithms.RSAOaep256;

export interface WrappedDataKey {
  readonly keyId: string;
  readonly wrappedKey: Uint8Array;
  readonly algorithm: typeof KEY_WRAP_ALGORITHM;
}

export interface EnvelopeDataKey extends WrappedDataKey {
  readonly plaintextKey: Buffer;
}

export class KeyManagement {
  public constructor(
    private readonly keyClient: KeyClient,
    private readonly credential: TokenCredential,
    private readonly keyName: string,
  ) {}

  public async generateAndWrapDataKey(): Promise<EnvelopeDataKey> {
    const plaintextKey = randomBytes(DATA_KEY_LENGTH_BYTES);

    try {
      const key = await this.keyClient.getKey(this.keyName);
      if (!key.id) {
        throw new Error(`Key Vault key "${this.keyName}" has no key ID`);
      }

      const cryptographyClient = new CryptographyClient(
        key.id,
        this.credential,
      );
      const result = await cryptographyClient.wrapKey(
        KEY_WRAP_ALGORITHM,
        plaintextKey,
      );

      return {
        plaintextKey,
        wrappedKey: result.result,
        keyId: key.id,
        algorithm: KEY_WRAP_ALGORITHM,
      };
    } catch (error) {
      plaintextKey.fill(0);
      throw new Error(
        `Failed to generate and wrap a data key with Key Vault key "${this.keyName}"`,
        { cause: error },
      );
    }
  }

  public async unwrapDataKey(
    keyId: string,
    wrappedKey: Uint8Array,
    algorithm: string,
  ): Promise<Buffer> {
    if (algorithm !== KEY_WRAP_ALGORITHM) {
      throw new Error(`Unsupported key-wrap algorithm: ${algorithm}`);
    }

    try {
      const cryptographyClient = new CryptographyClient(
        keyId,
        this.credential,
      );
      const result = await cryptographyClient.unwrapKey(
        KEY_WRAP_ALGORITHM,
        wrappedKey,
      );
      const plaintextKey = Buffer.from(
        result.result.buffer,
        result.result.byteOffset,
        result.result.byteLength,
      );

      if (plaintextKey.length !== DATA_KEY_LENGTH_BYTES) {
        plaintextKey.fill(0);
        throw new Error(
          `Unwrapped data key has ${plaintextKey.length} bytes; expected ${DATA_KEY_LENGTH_BYTES}`,
        );
      }

      return plaintextKey;
    } catch (error) {
      throw new Error(`Failed to unwrap the data key with Key Vault`, {
        cause: error,
      });
    }
  }
}

File: rerun-reports/02-python/20260829-032449/results/storage/data-plane/python/crud/storage-dp-python-encrypted-uploader/python-azure-skills/baseline/generated-code/key_management.py
from __future__ import annotations

import os
from dataclasses import dataclass

from azure.core.exceptions import HttpResponseError
from azure.keyvault.keys import KeyClient
from azure.keyvault.keys.aio import KeyClient as AsyncKeyClient
from azure.keyvault.keys.crypto import CryptographyClient, KeyWrapAlgorithm
from azure.keyvault.keys.crypto.aio import CryptographyClient as AsyncCryptographyClient
from azure.core.credentials import TokenCredential
from azure.core.credentials_async import AsyncTokenCredential

DATA_KEY_SIZE_BYTES = 32
KEY_WRAP_ALGORITHM = KeyWrapAlgorithm.rsa_oaep_256


class KeyManagementError(RuntimeError):
    """Raised when a Key Vault key operation fails."""


@dataclass(frozen=True)
class WrappedDataKey:
    key_id: str
    encrypted_key: bytes
    algorithm: str = KEY_WRAP_ALGORITHM.value


class KeyManager:
    def __init__(
        self,
        key_client: KeyClient,
        credential: TokenCredential,
        key_name: str,
        key_version: str | None = None,
    ) -> None:
        self._key_client = key_client
        self._credential = credential
        self._key_name = key_name
        self._key_version = key_version

    def generate_and_wrap_data_key(self) -> tuple[bytes, WrappedDataKey]:
        data_key = os.urandom(DATA_KEY_SIZE_BYTES)
        try:
            key = self._key_client.get_key(self._key_name, self._key_version)
            if not key.id:
                raise KeyManagementError("Key Vault returned a key without an ID")
            crypto_client = CryptographyClient(key, credential=self._credential)
            result = crypto_client.wrap_key(KEY_WRAP_ALGORITHM, data_key)
            return data_key, WrappedDataKey(
                key_id=key.id,
                encrypted_key=result.encrypted_key,
            )
        except HttpResponseError as exc:
            raise KeyManagementError(
                f"Key Vault could not wrap the data key with {self._key_name!r}: {exc}"
            ) from exc

    def unwrap_data_key(self, wrapped_key: WrappedDataKey) -> bytes:
        if wrapped_key.algorithm != KEY_WRAP_ALGORITHM.value:
            raise KeyManagementError(
                f"Unsupported key wrap algorithm: {wrapped_key.algorithm!r}"
            )
        try:
            crypto_client = CryptographyClient(
                wrapped_key.key_id,
                credential=self._credential,
            )
            result = crypto_client.unwrap_key(
                KEY_WRAP_ALGORITHM,
                wrapped_key.encrypted_key,
            )
            return result.key
        except HttpResponseError as exc:
            raise KeyManagementError(
                f"Key Vault could not unwrap the data key with "
                f"{wrapped_key.key_id!r}: {exc}"
            ) from exc


class AsyncKeyManager:
    def __init__(
        self,
        key_client: AsyncKeyClient,
        credential: AsyncTokenCredential,
        key_name: str,
        key_version: str | None = None,
    ) -> None:
        self._key_client = key_client
        self._credential = credential
        self._key_name = key_name
        self._key_version = key_version

    async def generate_and_wrap_data_key(self) -> tuple[bytes, WrappedDataKey]:
        data_key = os.urandom(DATA_KEY_SIZE_BYTES)
        try:
            key = await self._key_client.get_key(self._key_name, self._key_version)
            if not key.id:
                raise KeyManagementError("Key Vault returned a key without an ID")
            crypto_client = AsyncCryptographyClient(
                key,
                credential=self._credential,
            )
            result = await crypto_client.wrap_key(KEY_WRAP_ALGORITHM, data_key)
            return data_key, WrappedDataKey(
                key_id=key.id,
                encrypted_key=result.encrypted_key,
            )
        except HttpResponseError as exc:
            raise KeyManagementError(
                f"Key Vault could not wrap the data key with {self._key_name!r}: {exc}"
            ) from exc

    async def unwrap_data_key(self, wrapped_key: WrappedDataKey) -> bytes:
        if wrapped_key.algorithm != KEY_WRAP_ALGORITHM.value:
            raise KeyManagementError(
                f"Unsupported key wrap algorithm: {wrapped_key.algorithm!r}"
            )
        try:
            crypto_client = AsyncCryptographyClient(
                wrapped_key.key_id,
                credential=self._credential,
            )
            result = await crypto_client.unwrap_key(
                KEY_WRAP_ALGORITHM,
                wrapped_key.encrypted_key,
            )
            return result.key
        except HttpResponseError as exc:
            raise KeyManagementError(
                f"Key Vault could not unwrap the data key with "
                f"{wrapped_key.key_id!r}: {exc}"
            ) from exc
